From c7a8a4a96a6b15d2b500a13485c30abe72dbe508 Mon Sep 17 00:00:00 2001
From: Ramya Gnanasekar <quic_rgnanase@quicinc.com>
Date: Mon, 28 Nov 2022 21:01:21 +0530
Subject: [PATCH] cfg80211: Skip scan entry creation for duplicate beacon

An AP whose primary channel is non-PSC channel may transmit
duplicate beacon in PSC channel. To determine duplicate beacon
HE operation IE was used.
When validating the frequency, determined frequency was in KHZ
and chandef frequency was on MHZ resulting in false positives.
Also maximum operating bandwidth is 320MHz in 6GHz, hence difference
between operating frequency and PSC frequency can be more than 80Mhz.
Changed the condtion to meet the same.

Signed-off-by: Ramya Gnanasekar <quic_rgnanase@quicinc.com>

--- a/net/wireless/scan.c
+++ b/net/wireless/scan.c
@@ -1957,7 +1957,9 @@ cfg80211_get_bss_channel(struct wiphy *w
 	 * Frame info (beacon/prob res) is the same as received channel,
 	 * no need for further processing.
 	 */
-	if (freq == ieee80211_channel_to_khz(channel))
+	if (channel->band == NL80211_BAND_6GHZ &&
+	   (KHZ_TO_MHZ(freq) == channel->center_freq ||
+	    abs(KHZ_TO_MHZ(freq) - channel->center_freq) > 320))
 		return channel;
 
 	alt_channel = ieee80211_get_channel_khz(wiphy, freq);
